### PR TITLE
Movie testing

### LINKED ISSUE #10: Sprint 1 - Testing
- [ ] Establish testing suite
- [ ] test movies retrieved

### PR DESCRIPTION
**Pull request by ericberglund117**
## Description

- Fixed the bug that was displaying the movies multiple times on the homepage
- Created Movies.test file 
- Created apiCalls.js for all of our network requests
- Allow Movies component to access fetch request from apiCalls
- Update react testing library
- Update Movies.test file to be asynchronous 

## Motivation and Context

- We moved network request in a separate file in order to isolate the fetch call 
- This allowed us to mock the fetch request in our Movies.test.js.
- We also updated the react testing library so that 'waitFor' functionality was available. Was not in 9.x.x version of library.

## How Has This Been Tested?

- This has been tested through the browser 
- Through the Movies.test.js file 
- Test to see if Movies component was rendering movies 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ericberglund117** [comment]
@letakeane and @khalidwilliams

File: src/App.js [hunk 1]
```diff
@@ -4,15 +4,15 @@ import Film from './images/film-reel.png';
 import Tomatillos from './images/tomatillo.png';
 import Movies from './movies/Movies.js';
 import Login from "./login/Login.js";
-import { Switch, Route, Link } from 'react-router-dom';
+import { BrowserRouter, Switch, Route, Link } from 'react-router-dom';
```
Comment (MichaelEWalker87) [inline src/App.js:7]: Good use of getting a little head and using the BrowserRouter

File: src/App.js [hunk 2]
```diff
@@ -36,29 +36,35 @@ class App extends Component {
...
+              onClick={event => this.submitLogout(event)}>Logout</button>
+            <img className='flim-reel' src={Film}/>
+            <Login setUser={this.setUser} />
+          </header>
+            <Switch>
+              <Route exact path='/signin' render={() => <Login setUser={this.setUser} /> } />
+              <Route path='/' Component = <Movies /> />
+              // <Route path='/' render={() => <Movies movies={this.state.movies} /> } />
+              // <Route path='/logout' render={() => <App />} />
+            </Switch>
```
Comment (MichaelEWalker87) [inline src/App.js:60]: I am not sure in the route path are still connected but we can take a look

File: src/apiCalls.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
+export const getAllMovies = () => {
+  return fetch("https://rancid-tomatillos.herokuapp.com/api/v2/movies")
+  .then(response => response.json())
+};
```
Comment (MichaelEWalker87) [inline src/apiCalls.js:5]: Nice! I already started to use this file for the single movies

File: src/movies/Movies.js [hunk 4]
```diff
@@ -1,39 +1,37 @@
...
   constructor() {
     super();
     this.state = {
-      movies:[],
-    }
+      movies: [],
+    };
   }
 
-  getMovies = () => {
```
Comment (MichaelEWalker87) [inline src/movies/Movies.js:12]: it was good that we changed this to a componentDidMount

File: src/movies/Movies.test.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+       },
+    ]})
+    //render component
+    const { getByTitle, getByText, getByAltText } = render(<Movies  />);
+    // check that there is a container element on the page
+    const moviesContainer = getByTitle("movies-list");
+    const movieTitle = await waitFor(() => getByText('Money Plane'))
+    const movieDate = await waitFor(() => getByText('Release Date: 2020-09-29'))
+    const movieRating = await waitFor(() => getByText('Average Rating: 9'))
+    const moviePoster = await waitFor(() => getByAltText('image-poster'))
```
Comment (MichaelEWalker87) [inline src/movies/Movies.test.js:28]: great work on getting the await for async up

**Review by MichaelEWalker87 — APPROVED** [review]
I think we could use a little linting as well as cleaning up the busted code however I think so far it looks nice

File: package.json [hunk 6]
```diff
@@ -14,7 +12,7 @@
   "scripts": {
     "start": "react-scripts start",
     "build": "react-scripts build",
-    "test": "react-scripts test",
+    "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
```
Comment (khalidwilliams) [inline package.json:15]: Good catch, this is a maddening bug,

File: src/App.js [hunk 7]
```diff
@@ -36,29 +36,35 @@ class App extends Component {
...
     return (
       <section>
-        <header>
-          <Link to='/'>
-          <h1><img className='logo' src={Tomatillos}/>Rancid Tomatillos</h1>
-          </Link>
-          <h2> Welcome {this.state.user.name || 'Movie Goer'}!</h2>
-            <Link to='/signin'>
-            <button type="button">Login</button>
+      <BrowserRouter>
```
Comment (khalidwilliams) [inline src/App.js:39]: I would suggest rendering the BrowserRouter in the `index.js` file as a wrapper for the whole App. It gives us more flexibility when testing, and is a bit cleaner overall.

File: src/App.js [hunk 8]
```diff
@@ -36,29 +36,35 @@ class App extends Component {
...
+        <Route>
```
Comment (khalidwilliams) [inline src/App.js:40]: What is this `Route` doing for you? I don't think you need it here.

File: src/App.js [hunk 9]
```diff
@@ -36,29 +36,35 @@ class App extends Component {
...
-        // </section>
+            <h2> Welcome {this.state.user.name || 'Movie Goer'}!</h2>
+              <Link to='/signin'>
+              <button type="button">Login</button>
+              </Link>
+            <button
+              type="button"
+              onClick={event => this.submitLogout(event)}>Logout</button>
+            <img className='flim-reel' src={Film}/>
+            <Login setUser={this.setUser} />
```
Comment (khalidwilliams) [inline src/App.js:53]: Is the Login always being rendered? It seems like when you navigate to the `/signin` url you would see two `Login` components.

File: src/App.js [hunk 10]
```diff
@@ -36,29 +36,35 @@ class App extends Component {
...
+          </header>
+            <Switch>
+              <Route exact path='/signin' render={() => <Login setUser={this.setUser} /> } />
+              <Route path='/' Component = <Movies /> />
```
Comment (khalidwilliams) [inline src/App.js:57]: Don't forget to wrap expressions in `{ }`s

File: src/App.test.js [hunk 11]
```diff
@@ -1,9 +1,18 @@
...
-});
+// test('renders learn react link', () => {
+//   const { getByText } = render(<App />);
+//   const linkElement = getByText(/learn react/i);
+//   expect(linkElement).toBeInTheDocument();
+// });
+
+// describe('App', () => {
+//   it('should be able to render movies to the DOM', () => {
+//     const mockDisplayMovies = jest.fn();
```
Comment (khalidwilliams) [inline src/App.test.js:13]: We'll talk about how to mock out isolated functions tomorrow in class. You need to make sure that the version of the function that the App component will import when running this test is going to be a mock.

File: src/apiCalls.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+
+export const getAllMovies = () => {
+  return fetch("https://rancid-tomatillos.herokuapp.com/api/v2/movies")
+  .then(response => response.json())
```
Comment (khalidwilliams) [inline src/apiCalls.js:4]: Nice! Don't forget to add in some error handling for non-`2xx` level responses!

File: src/index.js [hunk 13]
```diff
@@ -6,11 +6,9 @@ import App from './App';
 import * as serviceWorker from './serviceWorker';
 
 ReactDOM.render(
-  <React.StrictMode>
     <BrowserRouter>
```
Comment (khalidwilliams) [inline src/index.js:9]: Oh nice! The `BrowserRouter` you're rendering inside of the App component is going to cause issues and you should need it if you're rendering one here.

File: src/movies/Movies.js [hunk 14]
```diff
@@ -1,39 +1,37 @@
...
+              <article key={index} className='poster-card' title='poster-card' id={movie.id}>
+                <img alt='image-poster' src={movie.poster_path}></img>
+                <section className='poster-card-text'>
+                  <h2 className='title-poster'>{movie.title}</h2>
+                  <h3 className='rating-poster'>Average Rating: {movie.average_rating}</h3>
+                  <h3 className='release-date-poster'>Release Date: {movie.release_date}</h3>
+                </section>
+              </article>)
+            })
+          }
```
Comment (khalidwilliams) [inline src/movies/Movies.js:33]: I wouild recommend taking this back out of the render statement and making it a variable or property that can be referred to, like it was before. We want the `render` method to be singularly responsible for returning UI, not iterating over data.

File: src/movies/Movies.test.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+         id: 694919,
+         poster_path: "https://image.tmdb.org/t/p/original//6CoRTJTmijhBLJTUNoVSUNxZMEI.jpg",
+         backdrop_path: "https://image.tmdb.org/t/p/original//pq0JSpwyT2URytdFG0euztQPAyR.jpg",
+         title: "Money Plane",
+         average_rating: 9,
+         release_date: "2020-09-29"
+       },
+    ]})
+    //render component
+    const { getByTitle, getByText, getByAltText } = render(<Movies  />);
```
Comment (khalidwilliams) [inline src/movies/Movies.test.js:22]: Try to opt for using `screen` for your queries

File: src/movies/Movies.test.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+import { render, waitFor, screen } from '@testing-library/react'
+import Movies from './Movies';
+import '@testing-library/jest-dom';
+import { getAllMovies } from '../apiCalls.js';
+jest.mock('../apiCalls.js');
+
+
+describe("Movies", () => {
+  it('should render a movie', async () => {
+    getAllMovies.mockResolvedValueOnce({ movies: [
```
Comment (khalidwilliams) [inline src/movies/Movies.test.js:11]: Nice job!! Make sure you can both explain what's happening here.

**Review by khalidwilliams — COMMENTED** [review]
Excellent job getting into async testing already! There are a few areas where you could tighten up your organization of the code, and a few spots where you need to rethink what you're doing with your router, but this is looking very good!

**Pull request merged**